### PR TITLE
No magazines mod

### PR DESCRIPTION
**Pull request by kevingranade**
Turns out this was pretty easy to do, so why not?

Reverts all guns to use integral magazines and blacklists all magazine items.
Integral magazine capacity is based on the default magazine for the gun.

### CODE REVIEW

**Comment by chaosvolt** [comment]
Ah, interesting to see.

File: src/item_factory.cpp [hunk 1]
```diff
@@ -94,26 +115,32 @@ void Item_factory::finialize_item_blacklist()
...
+}
+
+void Item_factory::load_item_option( JsonObject &json )
+{
+    add_to_set( item_options, json, "options" );
 }
 
 Item_factory::~Item_factory()
 {
+    item_options.clear();
```
Comment (BevapDin) [inline src/item_factory.cpp]: This should probably be in `Item_factory::clear`. The factory is only destroyed when the game is terminated, but `clear` is called before loading the game data.

Btw. why not make `item_options` a member or at least put it into an anonymous namespace/mark it as `static`?

File: src/item_factory.cpp [hunk 2]
```diff
@@ -66,10 +68,29 @@ void Item_factory::finialize_item_blacklist()
...
         }
     }
-    for (std::map<std::string, itype *>::const_iterator a = m_templates.begin(); a != m_templates.end();
-         ++a) {
+    const bool magazines_blacklisted = item_options.count("blacklist_magazines") != 0;
+    // Can't be part of the blacklist loop because the magazines might be
+    // deleted before the guns are processed.
+    if( magazines_blacklisted ) {
+        for( std::map<std::string, itype *>::const_iterator a = m_templates.begin();
+             a != m_templates.end(); ++a ) {
```
Comment (BevapDin) [inline src/item_factory.cpp]: Why not use a range-based loop? Same question for the loop below.

File: src/item_factory.cpp (hunk 1 above)
Comment (kevingranade) [inline src/item_factory.cpp]: Yes to both, thanks.

File: src/item_factory.cpp (hunk 2 above)
Comment (kevingranade) [inline src/item_factory.cpp]: Excuse my laziness, I was just copying, will fix soon.

File: src/item_factory.cpp [hunk 3]
```diff
@@ -66,10 +68,28 @@ void Item_factory::finialize_item_blacklist()
...
+    if( magazines_blacklisted ) {
+        for( std::pair<const std::string, itype *> &entry : m_templates ) {
+            itype *type = entry.second;
+            // find the guns, look up their default magazine, and add its capacity to the gun.
+            if( type->magazine_default.empty() ) {
+                continue;
+            }
+            itype *default_magazine = m_templates[ type->magazine_default.begin()->second ];
+            type->gun->clip = default_magazine->magazine->capacity;
+            type->gun->reload_time = default_magazine->magazine->reload_time;
```
Comment (mugling) [inline src/item_factory.cpp]: `reload_time` for magazines is per round so need to multiply by magazine `capacity`

**Comment by mugling** [comment]
> Turns out this was pretty easy to do, so why not?

Given you don't have to touch the reloading code to do this and we intend to continue supporting integral magazines in mainline this is an excellent solution.

> This might also render the old magazine mods useful again

The gunmod code is horrible and still has a lot of outstanding bugs. The required work to reimplement this isn't presently worth the outcome.

**Comment by chaosvolt** [comment]
> The gunmod code is horrible and still has a lot of outstanding bugs. The required work to reimplement this isn't presently worth the outcome.

Hmm. Understandable. If we are able to support the old magazine gunmods without additional hassle then it would be useful, but if proves to complicate things, then it's fine.

**Pull request merged**